### DISCUSSION TITLE
v0.15.0 (Webhook Notifications!)

### DISCUSSION THREAD

**Discussion by alexjustesen** [alexjustesen/speedtest-tracker#1107] [Releases]
## What's Changed
* [Chore] Added one time operations package by @alexjustesen in https://github.com/alexjustesen/speedtest-tracker/pull/1100
* [Feature] Add webhook notifications by @alexjustesen in https://github.com/alexjustesen/speedtest-tracker/pull/1103
* [Chore] Refactor notification channel tests by @alexjustesen in https://github.com/alexjustesen/speedtest-tracker/pull/1104
* [Chore] Refactored get jitter data method by @alexjustesen in https://github.com/alexjustesen/speedtest-tracker/pull/1105
* Release v0.15.0 by @alexjustesen in https://github.com/alexjustesen/speedtest-tracker/pull/1106

**Full Changelog**: https://github.com/alexjustesen/speedtest-tracker/compare/v0.14.7...v0.15.0

<hr /><em>This discussion was created from the release <a href='https://github.com/alexjustesen/speedtest-tracker/releases/tag/v0.15.0'>v0.15.0 (Webhook Notifications!)</a>.</em>

**Comment by masterwishx** [comment]
added https://discord.com/api/webhooks/XXXxXX/YYYY but not working , should i make some other steps ?

**Reply by alexjustesen** [reply]
Anything in the logs showing an error?

**Reply by masterwishx** [reply]
No errors, also no notification was sent, but in Gui it was sent...

**Reply by alexjustesen** [reply]
Did you try restarting the container and trying again?

**Reply by alexjustesen** [reply]
So looks like Discord requires some specific fields in the payload which is why it's rejecting the `POST` to the webhook. Current iterations of the webhook are meant for pretty basic payloads with just speedtest result info.

Going to say at this time Discord isn't supported but will make sure this gets added in the future.

https://discord.com/developers/docs/resources/webhook#execute-webhook

**Reply by masterwishx** [reply]
OK, waiting

**Comment by masterwishx** [comment]
![image](https://github.com/alexjustesen/speedtest-tracker/assets/28630321/10092b7d-83b8-4973-bca4-7412e2ff32f9)

![image](https://github.com/alexjustesen/speedtest-tracker/assets/28630321/3cc46112-8f09-41f0-b76d-16b49c4b1824)

**Comment by Starfiresg1** [comment]
Just noticed this - Webhook support is a great new feature.
Set it up with Gotify - test message worked out of the box 👍

**Reply by bonelifer** [reply]
@Starfiresg1 Could you give an example for Gotify.

**Reply by Starfiresg1** [reply]
Currently only the test message works - see https://github.com/alexjustesen/speedtest-tracker/discussions/1213

**Comment by publi0** [comment]
Hi there! I'm seeking clarification on whether this behavior is correct. 

When the webhook triggers and only one threshold has failed, the payload sent resembles this JSON structure:

```json
{
  "result_id": 1633,
  "site_name": "Speedtest Tracker",
  "metrics": [
    {
      "name": "Download",
      "threshold": "450 Mbps",
      "value": "447.69 Mbps"
    }
  ]
}
```

In this case, the `metrics` field is an array. However, when there are multiple thresholds, the payload changes to this format:

```json
{
  "result_id": 3931,
  "site_name": "Speedtest Tracker",
  "metrics": {
    "1": {
      "name": "Upload",
      "threshold": "400 Mbps",
      "value": "288.76 Mbps"
    },
    "2": {
      "name": "Ping",
      "threshold": "3 ms",
      "value": "4.41 ms"
    }
  }
}
```

Here, the `metrics` field is no longer an array; instead, it's represented as a map.

Context: I was implementing a listener for this webhook, and these JSON variations drew my attention.

**Reply by alexjustesen** [reply]
Great question, I'm open to feedback on what the format should be. I don't have any skin in the game for array vs. object map.

**Reply by publi0** [reply]
I find the map structure more advantageous, primarily because it simplifies the process of retrieving specific items. Using a descriptive key, such as the metric name, enhances clarity and ease of access:

```json
{
  "result_id": 3931,
  "site_name": "Speedtest Tracker",
  "metrics": {
    "upload": {
      "threshold": "400 Mbps",
      "value": "288.76 Mbps"
    },
    "ping": {
      "threshold": "3 ms",
      "value": "4.41 ms"
    }
  }
}
```

This approach eliminates the need to iterate through each item to locate the desired one. However, transitioning to this structure would necessitate a change in the current API contract. Should this pose an issue, I think reverting to the array version would be better. The current map implementation, with numeric keys, adds no additional information; it merely replicates the array index.
